### PR TITLE
DM-39466: Added flagging of sources near the INJECTED and INJECTED_TEMPLATE mask plane

### PR DESCRIPTION
**Pull request by BrunoSanchez**

### CODE REVIEW

File: python/lsst/ip/diffim/subtractImages.py [hunk 1]
```diff
@@ -651,18 +651,18 @@ def updateMasks(self, template, science, difference):
         mask = difference.mask
         mask &= ~(mask.getPlaneBitMask("DETECTED") | mask.getPlaneBitMask("DETECTED_NEGATIVE"))
 
+        self.log.info("Adding injected mask planes")
+        mask.addMaskPlane("INJECTED")
+        diffInjectedBitMask = mask.getPlaneBitMask("INJECTED")
+
+        mask.addMaskPlane("INJECTED_TEMPLATE")
+        diffInjTmpltBitMask = mask.getPlaneBitMask("INJECTED_TEMPLATE")
```
Comment (BrunoSanchez) [inline python/lsst/ip/diffim/subtractImages.py:659]: Move this into the if statement block. Same in 656.

File: python/lsst/ip/diffim/detectAndMeasure.py [hunk 2]
```diff
@@ -174,8 +174,14 @@ def setDefaults(self):
         self.forcedMeasurement.slots.shape = None
 
         # Keep track of which footprints contain streaks
-        self.measurement.plugins['base_PixelFlags'].masksFpAnywhere = ['STREAK']
-        self.measurement.plugins['base_PixelFlags'].masksFpCenter = ['STREAK']
+        self.measurement.plugins["base_PixelFlags"].masksFpAnywhere = ["STREAK"]
```
Comment (BrunoSanchez) [inline python/lsst/ip/diffim/detectAndMeasure.py:177]: Instead of append, just set the maskFpAnyWhere to the correct list.

**Review by isullivan — APPROVED** [review]
The code changes look fine, and my local tests with `ap_verify` pass. Please add a new unit test to check that `detectAndMeasure` correctly flags sources with footprints including pixels with the `INJECTED` or `INJECTED_TEMPLATE` mask set. A simple Jenkins re-run (without `ci_hsc` or `ci_imsim`) should be sufficient if you just add a test.

**Comment by BrunoSanchez** [comment]
Succesful jenkins run: 
https://rubin-ci.slac.stanford.edu/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/479/pipeline/

File: tests/test_detectAndMeasure.py [hunk 3]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+        science_fake_img, science_fake_sources = makeTestImage(
+            psfSize=2.4, xSize=xSize, ySize=ySize, seed=5, nSrc=3, noiseLevel=0.25, fluxRange=1
+        )
+        template, _ = makeTestImage(psfSize=2.4, xSize=xSize, ySize=ySize, doApplyCalibration=True)
+        tmplt_fake_img, tmplt_fake_sources = makeTestImage(
+            psfSize=2.4, xSize=xSize, ySize=ySize, seed=9, nSrc=3, noiseLevel=0.25, fluxRange=1
+        )
+        # created fakes and added them to the images
+        science.image.array += science_fake_img.image.array
+        template.image.array += tmplt_fake_img.image.array
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:450]: You can drop the `.array` in these two lines

File: tests/test_detectAndMeasure.py [hunk 4]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+        # adding mask planes to both science and template images
+        science.mask.addMaskPlane("FAKE")
+        science_fake_bitmask = science.mask.getPlaneBitMask("FAKE")
+        template.mask.addMaskPlane("FAKE")
+        template_fake_bitmask = template.mask.getPlaneBitMask("FAKE")
+
+        for a_science_source in science_fake_sources:
+            # 3 x 3 masking of the source locations is fine
+            bbox = lsst.geom.Box2I(
+                lsst.geom.Point2I(a_science_source.getX(), a_science_source.getY()), lsst.geom.Extent2I(3, 3)
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:462]: Perhaps use `bbox = a_science_source.getFootprint().getBBox()` instead.

File: tests/test_detectAndMeasure.py [hunk 5]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+            )
+            science[bbox].mask.array |= science_fake_bitmask
+
+        for a_template_source in tmplt_fake_sources:
+            # 3 x 3 masking of the source locations is fine
+            bbox = lsst.geom.Box2I(
+                lsst.geom.Point2I(a_template_source.getX(), a_template_source.getY()),
+                lsst.geom.Extent2I(3, 3)
+            )
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:471]: Perhaps `bbox = a_template_source.getFootprint().getBBox()` instead.

File: tests/test_detectAndMeasure.py [hunk 6]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+        # check subtraction mask plane is set where we set the previous masks
+        diff_mask = subtraction.difference.mask
+
+        # science mask should be now in INJECTED
+        inj_masked = (diff_mask.array & diff_mask.getPlaneBitMask("INJECTED")) > 0
+
+        # template mask should be now in INJECTED_TEMPLATE
+        injTmplt_masked = (diff_mask.array & diff_mask.getPlaneBitMask("INJECTED_TEMPLATE")) > 0
+
+        self.assertEqual(np.sum(inj_masked.astype(int)-science_fake_masked.astype(int)), 0)
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:490]: LSST developer guide requires spaces around `+` and `-` operators

File: tests/test_detectAndMeasure.py [hunk 7]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+        self.assertEqual(np.sum(injTmplt_masked.astype(int)-template_fake_masked.astype(int)), 0)
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:491]: For these two asserts, use `self.assertMasksEqual(inj_masked.astype(int), science_fake_masked.astype(int))`, which will give a more helpful error message if they are different.
I don't think there is an assert that checks arrays of booleans, so I think you do have to still do the `.astype(int)`. The tests pass without it, but if the arrays are actually different then the assert itself causes an error, which is unhelpful!
Note that in this case you could use `self.assertMasksEqual ` as I suggest, or `self.assertImagesEqual` or `assertFloatsEqual`

File: tests/test_detectAndMeasure.py [hunk 8]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+
+        # Now check that detection of fakes have the correct flag for injections
+        detectionTask = self._setup_detection()
+        excludeMaskPlanes = detectionTask.config.detection.excludeMaskPlanes
+        nBad = len(excludeMaskPlanes)
+        self.assertGreater(nBad, 0)
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:497]: Maybe make this more strict, with `self.assertEqual(nBad, 1)`

File: tests/test_detectAndMeasure.py [hunk 9]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+                                   subtraction.matchedTemplate,
+                                   subtraction.difference)
+
+        sci_refIds = []
+        tmpl_refIds = []
+        for diaSrc in output.diaSources:
+            if diaSrc['base_PsfFlux_instFlux'] > 0:
+                self._check_diaSource(science_fake_sources, diaSrc, scale=1, refIds=sci_refIds)
+                self.assertTrue(diaSrc['base_PixelFlags_flag_injected'])
+                self.assertTrue(diaSrc['base_PixelFlags_flag_injectedCenter'])
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:509]: Add:
```
                self.assertFalse(diaSrc['base_PixelFlags_flag_injected_template'])
                self.assertFalse(diaSrc['base_PixelFlags_flag_injected_templateCenter'])
```

File: tests/test_detectAndMeasure.py [hunk 10]
```diff
@@ -431,6 +431,87 @@ def _detection_wrapper(setFlags=True):
...
+            else:
+                self._check_diaSource(tmplt_fake_sources, diaSrc, scale=-1, refIds=tmpl_refIds)
```
Comment (isullivan) [inline tests/test_detectAndMeasure.py:511]: Add
```
                self.assertFalse(diaSrc['base_PixelFlags_flag_injected'])
                self.assertFalse(diaSrc['base_PixelFlags_flag_injectedCenter'])
```

**Review by isullivan — APPROVED** [review]
I've added my review of the new unit test

File: tests/test_detectAndMeasure.py (hunk 7 above)
Comment (BrunoSanchez) [reply tests/test_detectAndMeasure.py:491]: Chose `assertFloatsEqual` in this case. 
I have followed your suggestion. Thanks

File: tests/test_detectAndMeasure.py (hunk 5 above)
Comment (BrunoSanchez) [reply tests/test_detectAndMeasure.py:471]: Done

**Comment by BrunoSanchez** [comment]
I have applied the changes requested for the test. This improves significantly the quality of this unit test. Thanks very much

**Pull request merged**